File: cpp/36.cpp
#include <cctype>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  bool squareValid(vector<vector<char>> &board, int x, int y) {
    vector<int> freq(10, 0);
    for (int i = y - 1; i <= y + 1; i++)
      for (int j = x - 1; j <= x + 1; j++)
        if (isdigit(board[i][j]) && ++freq[board[i][j] - '0'] > 1)
          return false;
    return true;
  }
  bool colValid(vector<vector<char>> &board, int x) {
    vector<int> freq(10, 0);
    for (int i = 0; i < 9; i++)
      if (isdigit(board[x][i]) && ++freq[board[x][i] - '0'] > 1)
        return false;

    return true;
  }
  bool rowValid(vector<vector<char>> &board, int y) {
    vector<int> freq(10, 0);
    for (int i = 0; i < 9; i++)
      if (isdigit(board[i][y]) && ++freq[board[i][y] - '0'] > 1)
        return false;

    return true;
  }
  bool isValidSudoku(vector<vector<char>> &board) {

    int x = 1, y = 1;
    for (int i = 0; i <= 2; i++)
      for (int j = 0; j <= 2; j++)
        if (!squareValid(board, x + (3 * i), y + (3 * j)))
          return false;
    for (int i = 0; i < 9; i++)
      if (!colValid(board, i))
        return false;
    for (int i = 0; i < 9; i++)
      if (!rowValid(board, i))
        return false;

    return true;
  }
};

File: go/36.go
package main

func validSquare(board [][]byte, x int, y int) bool {
	freq := make(map[byte]int)
	for i := y - 1; i <= y+1; i++ {
		for j := x - 1; j <= x+1; j++ {
			if board[i][j] != byte('.') {
				freq[board[i][j]]++
				if freq[board[i][j]] > 1 {
					return false
				}
			}
		}
	}
	return true
}

func validRow(board [][]byte, y int) bool {
	freq := make(map[byte]int)
	for i := 0; i < 9; i++ {
		if board[y][i] != byte('.') {
			freq[board[y][i]]++
			if freq[board[y][i]] > 1 {
				return false
			}
		}
	}
	return true
}

func validCol(board [][]byte, x int) bool {
	freq := make(map[byte]int)
	for i := 0; i < 9; i++ {
		if board[i][x] != byte('.') {
			freq[board[i][x]]++
			if freq[board[i][x]] > 1 {
				return false
			}
		}
	}
	return true
}

func isValidSudoku(board [][]byte) bool {
	x, y := 1, 1

	// for square
	for i := 0; i <= 2; i++ {
		for j := 0; j <= 2; j++ {
			if !validSquare(board, x+3*i, y+3*j) {
				return false
			}
		}
	}

	// for col & row
	for i := 0; i < 9; i++ {
		if !validCol(board, i) {
			return false
		}
	}
  for i := 0; i < 9; i++ {
		if !validRow(board, i) {
			return false
		}
	}

	return true
}
